luaui/Widgets/api_resource_spot_builder.lua

```diff
@@ -142,122 +142,212 @@ local function IsSpotOccupied(spot)
-						spGiveOrderToUnit(id, CMD_STOP, {}, CMD_OPT_RIGHT)
+		local constructor = constructorIds[id]
+		if constructor then
+			-- iterate over constructor options to see if it can make the chosen extractor
+			for _, buildingId in pairs(constructor.building) do
+				if -buildingId == chosenExtractor and extractors[chosenExtractor] then
+					-- found match
+					local x, _, z = spGetUnitPosition(id)
+					if z then
+						ux, uz = ux+x, uz+z
+						latestMainBuilder = id
+						mainBuilders[#mainBuilders + 1] = id
+						if justDraw then
+							break -- prevent complex calculations further down the line
+						end
 					end
-					spGiveOrderToUnit(id, CMD_GUARD, { lastprocessedbestconstructor }, { "shift" })
+				else
+					secondaryBuilders[#secondaryBuilders + 1] = id
 				end
 			end
+
 		end
 	end
 
-	if mainBuildersCount == 0 then return end
-	aveX, aveZ = ux/mainBuildersCount, uz/mainBuildersCount
+	-- order secondary builders to guard main builders, equally dispersed
+	local index = 1
+	for i, uid in pairs(secondaryBuilders) do
+		if not options.shift then
+			spGiveOrderToUnit(uid, CMD_STOP, {}, CMD_OPT_RIGHT)
+		end
+		local mainBuilderId = mainBuilders[index]
+		spGiveOrderToUnit(uid, CMD_GUARD, { mainBuilderId }, { "shift" })
+		index = index + 1
+		if index > #mainBuilders then index = 1 end
+	end
+
+	if #mainBuilders == 0 then return end
+	aveX, aveZ = ux/#mainBuilders, uz/#mainBuilders
 
 	-- Get available mex spots within area
 	local commands = {}
-	local commandsCount = 0
 	local mexes = isGuard and { isGuard } or spots -- only need the mex/spot we guard if that is the case
 	for k = 1, #mexes do
 		local mex = mexes[k]
 		if not (mex.x % 16 == 8) then mexes[k].x = mexes[k].x + 8 - (mex.x % 16) end
 		if not (mex.z % 16 == 8) then mexes[k].z = mexes[k].z + 8 - (mex.z % 16) end
 		mex.x, mex.z = mexes[k].x, mexes[k].z
-		if Distance(cx, cz, mex.x, mex.z) < cr * cr then
-			if NoAlliedMex(mex.x, mex.z, maxresourceextractor) then
-				commandsCount = commandsCount + 1
-				commands[commandsCount] = { x = mex.x, z = mex.z, d = Distance(aveX, aveZ, mex.x, mex.z) }
+		if distance2dSquared(cx, cz, mex.x, mex.z) < cr * cr then
+			-- Skip mex spots that have queued mexes already
+			-- only searches selected builders, and only checks when shift is held
+			if options.shift then
+				if not resourceSpotHasExistingExtractorCommand(mex.x, mex.z, mainBuilders) then
+					if canExtractorBeUpgraded(mex.x, mex.z, chosenExtractor) then
+						commands[#commands + 1] = { x = mex.x, z = mex.z, d = distance2dSquared(aveX, aveZ, mex.x, mex.z) }
+					end
+				end
+			else
+				if canExtractorBeUpgraded(mex.x, mex.z, chosenExtractor) then
+					commands[#commands + 1] = { x = mex.x, z = mex.z, d = distance2dSquared(aveX, aveZ, mex.x, mex.z) }
+				end
 			end
 		end
 	end
 
 	-- Order available mex spots based on distance
 	local orderedCommands = {}
-	while commandsCount > 0 do
+	while #commands > 0 do
 		tasort(commands, function(a, b)
```

Comment:
Out of the scope of the PR most likely, but localizing the passed function is around [11x faster](https://springrts.com/wiki/Lua_Performance#TEST_8:_Functions_As_Param_For_Other_Functions).

It's a loop on commands only, should perhaps be reasonable.

Reply:
Good to know, thanks